File: app/src/main/java/com/limelight/preferences/PerfOverlayDisplayItemsPreference.java
package com.limelight.preferences;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.MultiSelectListPreference;
import android.preference.PreferenceManager;
import android.util.AttributeSet;

import com.limelight.R;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class PerfOverlayDisplayItemsPreference extends MultiSelectListPreference {
    
    private static final String DEFAULT_ITEMS = "resolution,decoder,render_fps,network_latency,decode_latency,host_latency,packet_loss,battery";
    
    public PerfOverlayDisplayItemsPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initialize();
    }

    public PerfOverlayDisplayItemsPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initialize();
    }

    public PerfOverlayDisplayItemsPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        initialize();
    }

    public PerfOverlayDisplayItemsPreference(Context context) {
        super(context);
        initialize();
    }

    private void initialize() {
        setEntries(R.array.perf_overlay_display_items_names);
        setEntryValues(R.array.perf_overlay_display_items_values);
        
        // 设置默认值（所有项目都默认选中）
        String[] defaultValues = DEFAULT_ITEMS.split(",");
        Set<String> defaultSet = new HashSet<>(Arrays.asList(defaultValues));
        setDefaultValue(defaultSet);
    }

    @Override
    protected void onAttachedToHierarchy(PreferenceManager preferenceManager) {
        super.onAttachedToHierarchy(preferenceManager);
        
        // 如果没有保存的值，设置默认值
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
        if (!prefs.contains(getKey())) {
            String[] defaultValues = DEFAULT_ITEMS.split(",");
            Set<String> defaultSet = new HashSet<>(Arrays.asList(defaultValues));
            setValues(defaultSet);
        }
    }
    
    /**
     * 获取默认的显示项目
     */
    public static Set<String> getDefaultDisplayItems() {
        String[] defaultValues = DEFAULT_ITEMS.split(",");
        return new HashSet<>(Arrays.asList(defaultValues));
    }
    
    /**
     * 检查特定项目是否被选中显示
     */
    public static boolean isItemEnabled(Context context, String itemKey) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        Set<String> selectedItems = prefs.getStringSet("perf_overlay_display_items", getDefaultDisplayItems());
        return selectedItems.contains(itemKey);
    }
    
    /**
     * 测试用：获取当前选中的所有显示项目
     */
    public static Set<String> getSelectedItems(Context context) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getStringSet("perf_overlay_display_items", getDefaultDisplayItems());
    }
    
    /**
     * 测试用：手动设置显示项目
     */
    public static void setDisplayItems(Context context, Set<String> items) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().putStringSet("perf_overlay_display_items", items).apply();
    }
} 

File: app/src/main/java/com/limelight/preferences/PerfOverlayDisplayItemsPreference.kt
package com.limelight.preferences

import android.content.Context
import android.util.AttributeSet

import androidx.preference.MultiSelectListPreference
import androidx.preference.PreferenceManager

import com.limelight.R
import com.limelight.utils.UiHelper

class PerfOverlayDisplayItemsPreference : MultiSelectListPreference {

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes) { initialize() }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int)
            : super(context, attrs, defStyleAttr) { initialize() }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) { initialize() }

    constructor(context: Context) : super(context) { initialize() }

    private fun initialize() {
        val nameEntries = context.resources.getTextArray(R.array.perf_overlay_display_items_names)
        val valueEntries = context.resources.getTextArray(R.array.perf_overlay_display_items_values)

        if (UiHelper.hasDisplayableBattery(context)) {
            entries = nameEntries
            entryValues = valueEntries
        } else {
            val filteredNames = ArrayList<CharSequence>()
            val filteredValues = ArrayList<CharSequence>()

            valueEntries.forEachIndexed { index, value ->
                if (value.toString() != BATTERY_ITEM) {
                    filteredNames.add(nameEntries[index])
                    filteredValues.add(value)
                }
            }

            entries = filteredNames.toTypedArray()
            entryValues = filteredValues.toTypedArray()
        }

        setDefaultValue(getDefaultDisplayItems(context))
    }

    override fun onAttachedToHierarchy(preferenceManager: PreferenceManager) {
        super.onAttachedToHierarchy(preferenceManager)

        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        if (!prefs.contains(key)) {
            values = getDefaultDisplayItems(context)
        } else if (!UiHelper.hasDisplayableBattery(context)) {
            val selectedItems = prefs.getStringSet(key, emptySet()) ?: emptySet()
            if (selectedItems.contains(BATTERY_ITEM)) {
                val filteredItems = selectedItems.filterNot { it == BATTERY_ITEM }.toSet()
                prefs.edit().putStringSet(key, filteredItems).apply()
                values = filteredItems
            }
        }
    }

    companion object {
        private const val BATTERY_ITEM = "battery"
        private const val DEFAULT_ITEMS = "resolution,decoder,render_fps,network_latency,decode_latency,host_latency,packet_loss,battery"

        fun getDefaultDisplayItems(context: Context): Set<String> {
            val defaultItems = DEFAULT_ITEMS.split(",").toMutableSet()
            if (!UiHelper.hasDisplayableBattery(context)) {
                defaultItems.remove(BATTERY_ITEM)
            }
            return defaultItems
        }

        fun isItemEnabled(context: Context, itemKey: String): Boolean {
            if (itemKey == BATTERY_ITEM && !UiHelper.hasDisplayableBattery(context)) {
                return false
            }

            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            val selectedItems = prefs.getStringSet("perf_overlay_display_items", getDefaultDisplayItems(context))
            return selectedItems?.contains(itemKey) == true
        }

        fun getSelectedItems(context: Context): Set<String>? {
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            return prefs.getStringSet("perf_overlay_display_items", getDefaultDisplayItems(context))
        }

        fun setDisplayItems(context: Context, items: Set<String>) {
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            prefs.edit().putStringSet("perf_overlay_display_items", items).apply()
        }
    }
}
